### PR TITLE
PERF: get_dummies create array of 0/1s with appropriate layout for internals

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This ensures that the 2D ndarray that is created in `get_dummies` for a single column is directly in the "right" layout as the internals prefer it (column-major, so it's row-major in the transposed Block.values).

For ArrayManager, this gives a speed-up for `get_dummies` itself (because we still copy the arrays to ensure proper contiguous memory layout for each column, currently):

```
In [7]: %timeit pd.get_dummies(s, sparse=False)
23.2 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- master
16.6 ms ± 689 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  <-- PR
```

For the current BlockManager, it makes the actual get_dummies a bit slower, but can make a follow-up operation actually faster (where a copy, or consolidation, or a reduction, or .. would happen):

```
In [4]: %timeit pd.get_dummies(s, sparse=False)
10.7 ms ± 397 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  <-- master
14.2 ms ± 713 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  <-- PR

In [5]: %timeit pd.get_dummies(s, sparse=False).copy()
22.3 ms ± 260 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- master
17.8 ms ± 1.92 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)  <-- PR
```

The timing is from the following example taken from the `frame_methods.py::GetDummies` ASV case:

```
import string

categories = list(string.ascii_letters[:12])
s = pd.Series(
    np.random.choice(categories, size=1000000),
    dtype=pd.CategoricalDtype(categories),
)
```

So it makes the actual `get_dummies` a bit slower for the BlockManager, and thus I could also explicitly only do this for the ArrayManager. But since it improves follow-up operations, I think it can actually be beneficial overall for BlockManager as well. But either is fine for me.

### CODE REVIEW

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -1014,7 +1014,8 @@ def get_empty_frame(data) -> DataFrame:
         return out
 
     else:
-        dummy_mat = np.eye(number_of_cols, dtype=dtype).take(codes, axis=0)
+        # take on axis=1 + transpose to ensure ndarray layout is column-major
```
Comment (jreback) [inline pandas/core/reshape/reshape.py:1017]: use the order paramater here: https://numpy.org/doc/stable/reference/generated/numpy.eye.html or maybe .reshape (as .take gives a copy)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/reshape/reshape.py:1017]: To clarify: the `take` is needed for the actual get_dummies operation. 
But given your suggestion, I thought that specifying `order` in `eye` might speed up the `take` call, but from a quick test that doesn't seem to matter

**Comment by jorisvandenbossche** [comment]
@jreback any further feedback here?

**Review by jreback — APPROVED** [review]

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (jreback) [reply pandas/core/reshape/reshape.py:1042]: guess this is fine

**Comment by jorisvandenbossche** [comment]
Thanks!

**Comment by simonjayhawkins** [comment]
> So it makes the actual `get_dummies` a bit slower for the BlockManager, and thus I could also explicitly only do this for the ArrayManager. But since it improves follow-up operations, I think it can actually be beneficial overall for BlockManager as well. But either is fine for me.

https://pandas.pydata.org/speed/pandas/#reshape.GetDummies.time_get_dummies_1d?python=3.8&Cython=0.29.21&commits=b8a71f33-154096f2

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -1039,7 +1039,8 @@ def get_empty_frame(data) -> DataFrame:
...
+        dummy_mat = np.eye(number_of_cols, dtype=dtype).take(codes, axis=1).T
```
Comment (jbrockmendel) [inline pandas/core/reshape/reshape.py:1043]: @jorisvandenbossche investigating a perf regression in `reshape.GetDummies.time_get_dummies_1d` current best guess is this line.  Would doing `arr.T.take(codes, axis=0)` be equivalent here?  That's about 7x faster for me locally.

Comment (jbrockmendel) [reply pandas/core/reshape/reshape.py:1043]: @jorisvandenbossche gentle ping; time_get_dummies_1d is the biggest perf regression im seeing vs 1.2.0

**Pull request merged**